commons/serialization/src/main/scala/sbt/pickling/json/JSONPickleFormat.scala

```diff
@@ -50,33 +53,123 @@ package json {
       }
     }
   }
+  object JSONPickleFormat {
+    private[json] val TYPE_TAG_FIELD = "$type"
+    private[json] val DYNAMIC_KEY_FIELD = "$keys"
+    private[json] val REF_ID_FIELD = "$ref"
+
+    private[json] def isSpecialField(name: String): Boolean =
+      (TYPE_TAG_FIELD == name) || (DYNAMIC_KEY_FIELD == name) || (REF_ID_FIELD == name)
+    private[json] def isElidedField(name: String): Boolean =
+      (DYNAMIC_KEY_FIELD == name)
+  }
 
-  class JSONPickleBuilder(format: JSONPickleFormat, buf: Output[String]) extends PBuilder with PickleTools {
-    private var nindent = 0
-    private def indent() = nindent += 1
-    private def unindent() = nindent -= 1
-    private var pendingIndent = false
-    private var lastIsBrace = false
-    private var lastIsBracket = false
-    private def append(s: String) = {
-      val sindent = if (pendingIndent) "  " * nindent else ""
-      buf.put(sindent + s)
-      pendingIndent = false
-      val trimmed = s.trim
-      if (trimmed.nonEmpty) {
-        val lastChar = trimmed.last
-        lastIsBrace = lastChar == '{'
-        lastIsBracket = lastChar == '['
+  sealed trait BuilderState {
+    def previous: BuilderState
+  }
```

Comment:
Assuming we continue with the state machine approach, would it make sense to hide the `var` transition?

```scala
def withState[A](f: BuilderState => (A, BuilderState)): A =
  {
    val (result, next) = f(state)
    state = next
    result
  }
```